code/modules/mecha/equipment/weapons/mecha_ballistic.dm

```diff
@@ -0,0 +1,233 @@
+// Enhanced mecha gun with projectile storage // THE SLOPPA
+/obj/item/mecha_parts/mecha_equipment/weapon/ballistic
+	name = "\improper General Ballistic Weapon"
+//	step_delay = 20 // It can't be THAT heavy!
+	var/max_projectiles = 0
+	var/projectiles = 0
+	var/projectile_energy_cost = 0
+	var/starts_full = FALSE
+	var/projectiles_cache = 0
+	var/projectiles_cache_max = 100
+	var/disabledreload = FALSE
+	var/list/ammo_types = list(/obj/item/ammo_casing/c9mm)
+	var/caliber = MM9
+	var/no_caliber = FALSE
+	var/list/loaded_projectiles = list()
+	var/projectile_type
+	var/burst_delay = 0 // Delay that seperates each projectile when burst-firing
+
+	var/projectiles_per_shot = 1
+	var/deviation = 0.7  //the shots were perfectly accurate no matter what this was set to
+
+/obj/item/mecha_parts/mecha_equipment/weapon/ballistic/action(atom/target)
+	if(!action_checks(target))
+		return
+	set_ready_state(0)
+	var/originaltarget = target
+	var/turf/curloc = get_turf(chassis)
+	var/turf/targloc = get_turf(target)
+
+	if(!curloc || !targloc)
+		do_after_cooldown()
+		return
+	var/start_loc = curloc
+	spawn
+		for(var/i=1 to min(projectiles, projectiles_per_shot))
+			if(!chassis)
+				break
+			var/turf/fire_from = start_loc
+			var/atom/current_target = originaltarget
+			var/turf/current_targloc = targloc
+			if(defective)
+				current_target = get_inaccuracy(originaltarget, 2, chassis)
+				current_targloc = get_turf(current_target)
+			if(!current_targloc || current_targloc == fire_from)
+				break
+			playsound(chassis, fire_sound, 80, 1)
+			var/obj/item/projectile/A = new projectile(fire_from)
+			src.projectiles--
+			A.firer = chassis.occupant
+			A.original = current_target
+			A.current = fire_from
+			A.starting = fire_from
+			A.yo = current_targloc.y - fire_from.y
+			A.xo = current_targloc.x - fire_from.x
+			A.OnFired()
+			spawn(0)
+				A.process()
+			if(i < min(projectiles, projectiles_per_shot))
+				sleep(burst_delay)
```

Comment:
Do not use sleep() unless there's absolutely no way around it. Since this resides within a proc on an /obj type, if this /obj is ever added to processing_objects for any reason then this single proc will delay the entire obj subsystem. Instead just do the following:

- Make a BURST_DELAY constant and assign it your delay value.
- Make a last_fired var for /obj/item/mecha_parts/mecha_equipment/weapon/ballistic.
- When the projectile fires, set last_fired to world.time.
- Before firing again, check that world.time > last_fired + BURST_DELAY.